File: 733-FloodFill/733-FloodFill.java
// Last updated: 8/16/2026, 9:49:36 PM
class Solution {
    // public int[][] floodFill(int[][] image, int sr, int sc, int color) {
    //     // connected components -> BFS
    //     if(image[sr][sc] == color) return image; // no color change required

    //     int m = image.length, n = image[0].length;
    //     int[][] res = image;
    //     int[][] directions = {{0,1}, {1,0}, {0,-1}, {-1,0}};
    //     int oldColor = image[sr][sc], newColor = color;

    //     Queue<int[]> queue = new LinkedList<int[]>();
    //     queue.offer(new int[]{sr, sc});
        
    //     while(!queue.isEmpty()) {
    //         int[] curr = queue.poll();
    //         res[curr[0]][curr[1]] = newColor;

    //         // iterate 4 directions
    //         for(int[] dir: directions) {
    //             int row = curr[0] + dir[0];
    //             int col = curr[1] + dir[1];
    //             // if [row, col] is valid (within matrix and value = oldColor)
    //             if(row >= 0 && row <= m-1 && col >= 0 && col <= n-1
    //                 && image[row][col] == oldColor) {
    //                     queue.offer(new int[]{row, col});
    //                 }
    //         }
    //     } 
    //     return res;
    // }

    // DFS recursion - faster than BFS
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int oldColor = image[sr][sc];
        if (oldColor == newColor) return image;
        
        int[][] res = image;
        dfs(sr, sc, image, res, oldColor, newColor);
        return res;
    }

    private void dfs(int row, int col, int[][] image, int[][] res, int oldColor, int newColor) {
        int m = image.length, n = image[0].length;
        if(row >= 0 && row <= m-1 && col >=0 && col <= n-1 
            && image[row][col] == oldColor) {
            res[row][col] = newColor;
            dfs(row - 1, col, image, res, oldColor, newColor);
            dfs(row, col - 1, image, res, oldColor, newColor);
            dfs(row + 1, col, image, res, oldColor, newColor);
            dfs(row, col + 1, image, res, oldColor, newColor);
        }

    }
}

File: 733-FloodFill/733-FloodFill.py
# Last updated: 8/16/2026, 9:49:30 PM
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        # # BFS state change, from curr_color -> color
        # curr_color = image[sr][sc]
        # if curr_color == color: return image

        # m, n = len(image), len(image[0])
        # dirs = [[0,1], [0,-1], [1,0], [-1,0]]
        
        # queue = deque()
        # queue.append([sr, sc])

        # while queue:
        #     curr = queue.popleft()
        #     # make the color change
        #     image[curr[0]][curr[1]] = color

        #     # find next nodes
        #     for dir in dirs:
        #         i = curr[0] + dir[0]
        #         j = curr[1] + dir[1]
        #         if i >= 0 and i <= m-1 and j >= 0 and j <= n-1 and image[i][j] == curr_color: 
        #             queue.append([i,j])
        # return image

        # DFS - recursion (slower)
        old_color = image[sr][sc]
        if old_color == color: return image

        self._helper(image, sr, sc, old_color, color)
        
        return image

    def _helper(self, image: List[List[int]], i: int, j: int, old_color: int, color: int) -> List[List[int]]:
        m, n = len(image), len(image[0])
        dirs = [[0,1], [0,-1], [1,0], [-1,0]]

        if i >= 0 and i <= m-1 and j >= 0 and j <= n-1 and image[i][j] == old_color: 
            # make the color (state) change
            image[i][j] = color 
            # iterate over next possibilities
            for dir in dirs:
                ii = i + dir[0]
                jj = j + dir[1]
                self._helper(image, ii, jj, old_color, color)

        return image